File: csharp/circular-buffer/CircularBuffer.cs
using System;

public class CircularBuffer<T>
{
    //TODO - compose Read and Write
    
    private readonly T[] _buffer;
    private readonly int _capacity;
    
    private int _head;
    private int _tail;
    private int _count;
    
    public CircularBuffer(int capacity)
    {
        _buffer = new T[capacity];
        _capacity = capacity;
        
        ResetContents();
    }

    public T Read()
    {
        if (_count == 0)
        {
            throw new InvalidOperationException();
        }

        T result = _buffer[_tail];
        _buffer[_tail] = default(T);

        
        _tail = IncrementOrWrap(_tail);

        _count--;
        
        return result;
    }

    public void Write(T value)
    {
        if (_count == _capacity)
        {
            throw new InvalidOperationException();
        }
        
        _buffer[_head] = value;
        _count++;

        _head = IncrementOrWrap(_head);
    }

    private int IncrementOrWrap(int index)
    {
        index++;
        if (index == _capacity) {
            index = 0;
        }

        return index;
    }


    public void Overwrite(T value)
    {
        _buffer[_tail] = value;
        _tail = IncrementOrWrap(_tail);
    }

    public void Clear() => ResetContents();

    private void ResetContents() {
        _head = 0;
        _tail = 0;
        _count = 0;
    }
}

File: java/circular-buffer/src/main/java/CircularBuffer.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

class CircularBuffer<T> {

    private int writeIndex;
    private  int readIndex;
    List<T> data;
    private int size;

    CircularBuffer(final int size) {
        this.size = size;
        clear();
    }

    T read() throws BufferIOException {
        if (readIndex == -1 || this.data.get(readIndex) == null) {
            throw new BufferIOException("Tried to read from empty buffer");
        }

        T returnVal = this.data.get(readIndex);
        this.data.set(readIndex, null);
        readIndex += 1;
        if (readIndex >= data.size()) {
            readIndex = 0;
        }
        return returnVal;
    }

    void write(T data) throws BufferIOException {
        if (readIndex == -1) {
            readIndex = 0;
        }
        if (this.data.get(writeIndex) != null) {
            throw new BufferIOException("Tried to write to full buffer");
        }
        this.data.set(writeIndex, data);
        writeIndex += 1;
        if (writeIndex >= this.data.size()) {
            writeIndex = 0;
        }
    }

    void overwrite(T data) {
        this.data.set(writeIndex, data);
        writeIndex += 1;
        if (writeIndex >= this.data.size()) {
            writeIndex = 0;
        }
    }

    void clear() {

        data = new LinkedList<>();
        for (int i = 0; i < this.size; i++) {
            data.add(null);
        }
        writeIndex = 0;
        readIndex = -1;}

}